types/bools/compare.go

```diff
@@ -0,0 +1,18 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package bools contains the bools.Compare function.
+package bools
+
+// Compare compares two boolean values as if false is ordered before true.
+func Compare[T ~bool](x, y T) int {
+	var false T
```

Comment:
Remove and compare `~bool` directly to the untyped bool `false`?

Reply:
The downside of directly operating on a `bool` is that it's more annoying to call with named bools.

As currently written, you can do:
```go
type MyBool bool
var a, b MyBool

bools.Compare(a, b)
```
Otherwise you would have to do:
```go
bools.Compare(bool(a), bool(b))
```

We don't have many named bools in our codebase but they exist.

Reply:
Yeah, I agree. Just saying that you probably don't need to redefine `false` as a typed variable; the standard `false` const is an untyped bool.

In other words, the below code compiles and prints "ok":
```go
var a MyBool
if a == false {
  fmt.Println("ok")
}
```

Reply:
Ah yes, thank you.

I arrived at the current code cause I was a bit annoying that you can't do `!x` if `T` is a `~bool`.

Reply:
Hmm.. TIL, and it appears this might be the reason:
[spec: switch with no value uses bool rather than untyped bool](https://github.com/golang/go/issues/6398)

Doing a switch on `T(true)` would work, but I'm unsure which one is better:
```go
func Compare[T ~bool](x, y T) int {
	switch T(true) {
	case !x && y:
		return -1
	case x && !y:
		return +1
	default:
		return 0
	}
}
```

Reply:
Surprisingly the current implementation is less compiled assembly:

With `x == false`:
```
    0x000e 00014 TESTB   AL, AL
    0x0010 00016 JNE 36
    0x0012 00018 TESTB   CL, CL
    0x0014 00020 JEQ 32
    0x0016 00022 MOVQ    $-1, AX
    0x001d 00029 JMP 49
    0x001f 00031 NOP
    0x0020 00032 TESTB   AL, AL
    0x0022 00034 JEQ 47
    0x0024 00036 TESTB   CL, CL
    0x0026 00038 JNE 47
    0x0028 00040 MOVL    $1, AX
    0x002d 00045 JMP 49
    0x002f 00047 XORL    AX, AX
    0x0031 00049 MOVQ    AX, tailscale.com/types/bools.sink(SB)
    0x0038 00056 RET
```
With `switch T(true)`
```
    0x000e 00014 TESTB   AL, AL
    0x0010 00016 JEQ 22
    0x0012 00018 XORL    DX, DX
    0x0014 00020 JMP 24
    0x0016 00022 MOVL    CX, DX
    0x0018 00024 TESTB   DL, DL
    0x001a 00026 JEQ 37
    0x001c 00028 MOVQ    $-1, AX
    0x0023 00035 JMP 61
    0x0025 00037 TESTB   AL, AL
    0x0027 00039 JEQ 46
    0x0029 00041 XORL    $1, CX
    0x002c 00044 JMP 48
    0x002e 00046 XORL    CX, CX
    0x0030 00048 TESTB   CL, CL
    0x0032 00050 JEQ 59
    0x0034 00052 MOVL    $1, AX
    0x0039 00057 JMP 61
    0x003b 00059 XORL    AX, AX
    0x003d 00061 MOVQ    AX, tailscale.com/types/bools.sink(SB)
    0x0044 00068 RET
```